components/br-stream/src/codec.rs

```diff
@@ -0,0 +1,54 @@
+// Copyright 2021 TiKV Project Authors. Licensed under Apache-2.0.
+use bytes::{Buf, Bytes};
+use std::io::prelude::*;
+use std::io::Cursor;
+
+pub struct Encoder;
+
+impl Encoder {
+    // TODO use a more efficent encode kv event format
+    pub fn encode_event(key: &[u8], value: &[u8]) -> Vec<u8> {
+        let mut buf = vec![];
+        let key_len = (key.len() as u32).to_be_bytes();
+        let val_len = (value.len() as u32).to_be_bytes();
+        buf.extend_from_slice(&key_len);
+        buf.extend_from_slice(key);
+        buf.extend_from_slice(&val_len);
+        buf.extend_from_slice(value);
+        buf
+    }
+
+    #[allow(dead_code)]
+    pub fn decode_event(e: &[u8]) -> (Vec<u8>, Vec<u8>) {
+        let mut buf = Cursor::new(Bytes::from(e));
+        let len = buf.get_u32_be() as usize;
+        let mut buffer = vec![0; len];
+        buf.read_exact(buffer.as_mut_slice()).unwrap();
+        let key = buffer.to_vec();
+        let len = buf.get_u32_be() as usize;
+        let mut buffer = vec![0; len];
+        buf.read_exact(buffer.as_mut_slice()).unwrap();
+        let val = buffer.to_vec();
```

Comment:
seems buffer is already `Vec`? (This `to_vec` may introduce unnecessary copying)